### ISSUE TITLE
Why do we need to have defaultFinalStep?

### ISSUE THREAD

**Issue by SergeStinckwich** [cormas/cormas#155]
There is a method ``CormasModel class >> defaultFinalStep``. He was returning nil until now, I change it to ^100 now.

But why do we need a defaultFinalStep for a simulation? Because this is very difficult to know this value.

**Comment by tomooda** [comment]
More precisely, it was subclassResponsibility and each subclass actually defines its own default value. And I agree with @SergeStinckwich that it's hard to predict an appropriate value.

For example, if you want to set the simulation steps to 150, you can either
* set the value in the simulation code, e.g. ```aModel simManager finalTime: 150```,
* define the ```defaultFinalStep``` method in your model class, or
* the ```defaultFinalStep``` method in the CormasModel class will be used. It is now set to 100.

In my opinion, ```aModel simManager finalTime: 150``` is very descriptive way.
If you want to omit it in a simulation code, you need to define a ```defaultFinalStep``` method in your model class.
Here is a problem.
CormasModel is already a very big class. It'd be hard to discover that you need to define a ```defaultFinalStep``` from tons of other methods defined in the CormasModel class.
If CormasModel class>>defaultFinalStep is subclass responsibility, it would be relatively easier to find it.
In return, you'll need to define a ```defaultFinalStep``` method everytime you create your model class, which is quite tedious.

My question to modelers is:

Do you usually set the number of simulation steps in your simulation code?

If yes, I'd suggest to set the ```CormasModel class>>defaultFinalStep``` to 100.
It's convenient. You don't have to define your model's ```defaultFinalStep``` method which is rarely used.

If no, I'd suggest to revert it to subclassResponsibility to encourage modelers to define ```defaultFinalStep``` in each model class.

**Comment by SergeStinckwich** [comment]
Thank you Tomo. This was my state of mind 👍 
At the beginning, defaultFinalStep was returning nil, so I change it as self subclassResponsability, but it enforces all the subclasses to define it ...

**Comment by pierbo** [comment]
I think this default value could be set to 100 steps (even if it can be 
redefined by subclasses).

defaultFinalStep
     ^ 100

Le 08/06/2020 à 07:38, Serge Stinckwich a écrit :

**Comment by pierbo** [comment]
Hi Tomo,

Le 08/06/2020 à 06:43, Tomohiro Oda a écrit :
> ...
>
> My question to modelers is:
>
> Do you usually set the number of simulation steps in your simulation code?
>
No !

In Cormas-VW, the attribute #|defaultFinalStep |doesn't exist.

As the management of time is the responsibility of SimManager, I would 
let it at this level ("lazy initialization"):

finalStep
     ^finalStep isNil
         ifTrue:
             [finalStep := 100 asValue]
         ifFalse:
             [finalStep]

When the user wants to run a complete simulation, he must specify the 
finalStep (in the simulation interface). Programmatically, it could be 
by setting the finalStep from a "delegation" method in CMCormasModel:

finalStep: x

     self simManager finalStep: x

So, for me, I don't see the interest of keeping |defaultFinalStep |as 
finalStep is set by the lazy initialization.

**Comment by SergeStinckwich** [comment]
@pierbo you don't answer the question 😃 
Why a modeller need to fix a final step value? This is completely arbitrary for me.

You are right, there is no defaultStep attribute, because this is a method in fact.

The action to do is not clear: remove defaultStep and use finalStep in the CMSimManager?

**Comment by pierbo** [comment]
Well, a simulation stops when step > finalStep. So to run a simulation, you must specify a finalStep.
But indeed, we can also say that if finalStep is nil, then the simulation would run forever.
We have to decide: 
   - define the finalStep to 100 (by default, but can be changed), or
   - if finalStep = nil, run forever (until user break)
I prefer the 1st option, but won't fight for that !!   :)

**Comment by ElCep** [comment]
@SergeStinckwich he's right, it's arbitrary, but so is running the simulation until break. Between those two shelters as @pierbo , I have no opinion. :smile_cat:

**Comment by SergeStinckwich** [comment]
We can remove this defaultFinalStep value, because normally you run a simulation step by step or continuously.

**Comment by tomooda** [comment]
I agree with removing CormasModel class>>defaultFinalStep and modify CMSimManager>>finalStep defaults to 100 if none specified. May I write a PR?

**Comment by SergeStinckwich** [comment]
ok let's wait answers from @pierbo

**Comment by ElCep** [comment]
@SergeStinckwich @tomooda like @pierbo  says, he's not gonna fight for it, so if you want to do PR, go ahead. :smile: :flight_departure:

**Comment by pierbo** [comment]
Yes, I think we should remove defaultFinalStep from CormasModel, and just keep in SimManager, the attribute #finalStep + lazy initialization:
finalStep
     ^finalStep ifNil:[finalStep := 100]

BTW I misspoke: it is not up to the modeler to define the finalStep, but to the user when he wants to run a simulation.

**Comment by ElCep** [comment]
@pierbo , what I understood from @SergeStinckwich conserne is to let the model running without final step by default. And Isn't that what you're proposing with this lazy init . 
So my question: in Netlogo or Gama, by default the model runs infinitely. Why historically in Cormas the choice was made to limit the simulation in time?

**Comment by pierbo** [comment]
Yes, I got it. From the creation of Cormas (1998!), they had chosen this option. I see no problem in deciding that if finalStep is nil, then the simulation runs endlessly. We can even keep the 2 options: by default, finalStep = 100, but if the user defines finalStep = nil then it is endless.
